### PR TITLE
Optimizer

### PR DESCRIPTION
**Pull request by malin1993ml**
This PR contains the Orca-style framework for query optimization. There're four stages of query optimization:
(1) logically equivalent operator exploration;
(2) physical operator implementation;
(3) cost operator expressions given required properties;
(4) choose the least cost operator tree and transform to executor plan tree.

Right now the optimizer can only optimize select statements with star, columns, and simple projection. Other operators and transformations left to be implemented. I also left some placeholders for them in the implementation.

~~**Note:** Please do not merge this PR yet. I created this PR now to check the coverage of the code with coverall. I may add additional tests and will notify when I'm done.~~

### CODE REVIEW

**Comment by malin1993ml** [comment]
I'm done with adding tests for now. Feel free to review the PR once travis is happy.

**Comment by apavlo** [comment]
@hzxa21 You should look at this PR. You can help out with the code review This is what you will work on next semester.

File: src/include/parser/create_statement.h [hunk 1]
```diff
@@ -162,7 +163,7 @@ struct CreateStatement : TableRefStatement {
       : TableRefStatement(STATEMENT_TYPE_CREATE),
         type(type),
         if_not_exists(false),
-        columns(NULL) {};
+        columns(NULL){};
```
Comment (apavlo) [inline src/include/parser/create_statement.h]: Switch to `nullptr`

File: test/optimizer/rule_test.cpp [hunk 2]
```diff
@@ -14,7 +14,7 @@
 
 #define private public
 
-#include "optimizer/op_expression.h"
+#include "../../src/include/optimizer/operator_expression.h"
```
Comment (apavlo) [inline test/optimizer/rule_test.cpp]: Does this need the `../../` path?

File: src/include/parser/create_statement.h (hunk 1 above)
Comment (malin1993ml) [inline src/include/parser/create_statement.h]: Fixed

File: test/optimizer/rule_test.cpp (hunk 2 above)
Comment (malin1993ml) [inline test/optimizer/rule_test.cpp]: That was because of the auto-refactoring of Eclipse. Already fixed.

**Pull request merged**